packages/core/src/shorthands/flex.ts

```diff
@@ -0,0 +1,163 @@
+
+/**
+ * generateFlex is the core functionality for expanding a `flex` shortcut into a longhand `flexGrow`, `flexShrink`,
+ * and `flexBasis`
+ * @param values shortcut to expand
+ * @returns FlexStyle longhand
+ */
+function generateFlex(...values: FlexInput): FlexStyle {
+  const isOneValueSyntax = values.length === 1;
+  const isTwoValueSyntax = values.length === 2;
+  const isThreeValueSyntax = values.length === 3;
+
+  if (isOneValueSyntax) {
+    const [firstValue] = values;
+
+    if (isInitial(firstValue)) {
+      return {
+        flexGrow: 0,
+        flexShrink: 1,
+        flexBasis: 'auto',
+      };
+    }
+
+    if (isAuto(firstValue)) {
+      return {
+        flexGrow: 1,
+        flexShrink: 1,
+        flexBasis: 'auto',
+      };
+    }
+
+    if (isNone(firstValue)) {
+      return {
+        flexGrow: 0,
+        flexShrink: 0,
+        flexBasis: 'auto',
+      };
+    }
+
+    if (isUnitless(firstValue)) {
+      return {
+        flexGrow: firstValue as number,
+        flexShrink: 1,
+        flexBasis: 0,
+      };
+    }
+
+    if (isWidth(firstValue)) {
+      return {
+        flexGrow: 1,
+        flexShrink: 1,
+        flexBasis: firstValue as Width,
+      };
+    }
+  }
+
+  if (isTwoValueSyntax) {
+    const [firstValue, secondValue] = values;
+
+    if (isUnitless(secondValue)) {
+      return {
+        flexGrow: firstValue,
+        flexShrink: secondValue,
+        flexBasis: 'auto',
+      };
+    }
+
+    if (isWidth(secondValue)) {
+      return {
+        flexGrow: firstValue,
+        flexShrink: 1,
+        flexBasis: secondValue as Width,
+      };
+    }
+  }
+
+  if (isThreeValueSyntax) {
+    const [firstValue, secondValue, thirdValue] = values;
+    if (isUnitless(firstValue) && isUnitless(secondValue) && (isAuto(thirdValue) || isWidth(thirdValue))) {
+      return {
+        flexGrow: firstValue,
+        flexShrink: secondValue,
+        flexBasis: thirdValue as Width,
+      };
+    }
+  }
+
+  return {
+    flexGrow: 0,
+    flexShrink: 1,
+    flexBasis: 'auto',
+  } as FlexStyle;
+}
+
+/**
+ * mergeValue assembles an array of fallbacks from unique values.
+ * @param to the accumulator
+ * @param value to assess
+ * @returns Either (1) a singular FlexStyle if all the values were repeated, (2) or an array of FlexStyle fallbacks. The
+ * values in the array are not repeated.
+ */
+const mergeValue = (to: ValueOrArray<CSS.Property.Flex>, value: CSS.Property.Flex): ValueOrArray<CSS.Property.Flex> => {
+  if (Array.isArray(to)) {
+    return to[to.length - 1] === value ? to : [...to, value];
+  } else {
+    return to === value ? to : [to, value];
+  }
+};
+
+/**
+ * mergeFallbacks takes an array of FlexStyles and merges then into one object where each key can have an array of
+ * fallbacks.
+ * @param fallbackArray
+ * @returns FlexStyle with fallback arrays if needed.
+ */
+const mergeFallbacks = (fallbackArray: FlexInput[]) =>
```

Comment:
Can you provide some explanation about the fallback ? I don't see any mention of fallbacks for flex in the mdn docs

Reply:
I was following [this API design (CSS Fallback Properties)](https://griffel.js.org/react/api/make-styles#css-fallback-properties) so `shorthands.flex` can allow the user to provide multiple values as fallbacks. Here is the usage I intended:

_Input_
```ts
flex([0, 0, 'auto'], [0, 0, '100vw'])
```

_Output_
```css
.a {
    flex-grow: 0;
    flex-shrink: 1;
    flex-basis: 'auto';
    flex-basis: '100vw';
}
```

If you don’t think it is necessary, I can remove it. It doesn’t look like other `shorthands` utilities have support for that. If that is the way you want to go, I will only keep `generateFlex` and rename it `flex`.

---

P.S. It would be good to add support for fallbacks though. Probably better though a generic abstraction that could be applied to all other “expanders”.

Reply:
As you noticed we have support for fallback values, but it's not a goal of shorthands functions to do this.

I suggest to limit it to form of `flex(0, 0, ['auto', '100vw'])`, it will simplify implementation and will be consistent with other functions.

Reply:
I've been pondering about this today… Letting the user pass arrays of values would make the expanding implementation actually more complex 🤔 With `flex` the type of the value and not the position defines what it is going to be when expanded. So if say someone passes `flex(1, ['10px', 2])` those should expand into:

```js
{
      flexGrow: 1,
      flexShrink: [1, 2],
      flexBasis: ['10px', 'auto'],
}
```
…because the second value could be both `flexShrink` and `flexBasis`. Same for say `['auto', 'none']`.

I can either add this logic _or_ restrict the usage of flex with three value syntax only. Although in the second case values like `flex('auto')`, flex('initial')`, etc. will fall through the cracks.

Personally, I would probably go for more complex logic for the sake of flexibility for the user, but will let you decide!

Reply:
I would just need to rewrite the `mergeValue` and `mergeFallbacks` helpers differenty and replace `flex` with `generateFlex`.

Reply:
I have an opposite idea then: let's remove fallbacks for now due their complexity from `flex()` implementation.
It will be a bit inconsistent, but solves complexity issues. The general advice will be then: _You know about fallbacks, it's a rare case ➡️ you can use CSS longhands for them._

Reply:
> you can use CSS longhands for them.

@layershifter we'd need to explicitly enable CSS longhand for `flex` (they are stripped currently AFAIR)
https://github.com/microsoft/griffel/blob/0e24a70c7229e71f5828161e5616f3fe72e63e03/packages/core/src/constants.ts#L58

Reply:
@ling1726 why? What cannot be done with CSS longhands?

Reply:
ah sorry you meant explicitly use properties like `flexBasis`, my bad

Reply:
@layershifter, yeah that makes sense 👌👌

Reply:
Done